### ISSUE TITLE
Communicate early, in the YZ stage

### ISSUE THREAD

**Issue by npadmana** [npadmana/DistributedFFT#23]
cc @ronawho -- any thoughts?

In the current version of the code, we do all our communication in the X stage, but this actually makes overlapping communication and computation difficult. A different approach would be to actually fire the communications in the YZ stage as follows:

```chpl 
// Do the y transforms first (we'll see why in a second)
forall {ix,iz} in {myxRange, zRange} {
  FFT1D_Strided(arr[ix,y0,iz]); // y is not contiguous
}
// Do the z transforms 
forall {ix,iy} in {myxRange, yRange} {
  FFT1D(arr[ix,iy, z0]);
  // At this point, the ix,iy skewer can be sent
  chpl_comm_put(dest[iy,ix,..],arr[ix,iy,..]); // This is a single contiguous put.
}
```
After this is done, the `X` transform can be done, but it'll be completely local.

Some questions for you
1. I'm assuming there is a `chpl_comm_put` primitive; what are its arguments?
2. Ideally, the PUT should be non-blocking, because it'll allow the thread to move on to the next operation, overlapping communication with the FFT computation. Is there a way to fake this?

**Comment by npadmana** [comment]
I'll note that doing the code this way works well for real-real and complex-complex transforms, but not real-complex and complex-real. But that's ok for now.

**Comment by ronawho** [comment]
Should be something like `__primitive("chpl_comm_put", srcRef, dstRef.locale.id, dstRef, size);`

**Comment by ronawho** [comment]
Internally the chpl_comm_put call will yield while waiting for the transfer to complete. You can get comm/compute overlap, by doubling up the number of tasks, but this may start to have bad cache effects (experiments I was doing with trying to make pencilFFT-2 non-blocking seemed to hurt performance)

Still, you can try this with something like:

```chpl
forall {ix,iy} in {myxRange, yRange}.these(tasksPerLocale=here.maxTaskPar*2) {
  FFT1D(arr[ix,iy, z0]);
  // At this point, the ix,iy skewer can be sent
  chpl_comm_put(dest[iy,ix,..],arr[ix,iy,..]); // This is a single contiguous put.
}
```

(I think this is one of the few internal iterators thas has tuning options mentioned in https://github.com/npadmana/DistributedFFT/pull/19#issuecomment-536029707)

**Comment by npadmana** [comment]
Could one wrap the PUT in a begin?

**Comment by ronawho** [comment]
Yeah, you could do something like:

```chpl
sync {
  forall {ix,iy} in {myxRange, yRange} {
    FFT1D(arr[ix,iy, z0]);
    // At this point, the ix,iy skewer can be sent
  begin {
    chpl_comm_put(dest[iy,ix,..],arr[ix,iy,..]); // This is a single contiguous put.
  }
}
```

**Comment by ronawho** [comment]
Simple copy routine to wrap the primitives:

```chpl
proc copy(ref dst, const ref src, size) {
  if dst.locale.id == here.id {
    __primitive("chpl_comm_get", dst, src.locale.id, src, size.safeCast(size_t));
  } else if src.locale.id == here.id {
    __primitive("chpl_comm_put", src, dst.locale.id, dst, size.safeCast(size_t));
  } else {
    halt("Remote src and remote dst not yet supported");
  }
}
```

Sample usage:
```chpl
config const n = 100000;
var arr1: [1..n] int = 1..n;
var arr2: [1..n] int;
on Locales[1] {
  var arr3: [1..n] int;
  copy(arr3[arr3.domain.first], arr1[arr1.domain.first], n*numBytes(int)); // GET
  copy(arr2[arr2.domain.first], arr3[arr3.domain.first], n*numBytes(int)); // PUT
  writeln((arr3[1], arr3[n]));
}
writeln((arr1[1], arr1[n]));
writeln((arr2[1], arr2[n]));
```

**Comment by npadmana** [comment]
@ronawho -- this looks promising. Take a look at branch `pencilFFT-2-comm-yz`.

For case D, on swan:

| nodes | pencilFFT-2 | comm-yz |
|---|---|---|
| 8 | 41.4 | 33.8 |
| 16 | 27.1 | 20.4 | 
| 32 | 15.4 | 13.1 |

So that looks good. Unfortunately, it segfaults when I try to run `E` -- the error seems to occur in an `avx_memcpy` in a `remote_get`. I haven't had a chance to track down if the bug is in my code or somewhere else. 

Thoughts?

**Comment by ronawho** [comment]
Cool. 

We're probably writing to invalid memory -- Maybe we're passing the wrong size or writing to the wrong starting point. I'd start by running without `--fast` and see if we catch any chapel level OOB or anything. That won't catch errors in the PUT primitive though. Maybe replace with slice assignment, see if it hits any OOB and if that works, then we're probably passing something wrong to the PUT. 

(I should have a few hours later tonight to look more if you don't beat me to it)

**Comment by npadmana** [comment]
Interestingly, I can't seem to reproduce the segfault... I've been able to run class E cases multiple times with no errors.

For case E, on Swan --

| nodes | pencilFFT-2 | comm-yz |
|---|---|---|
| 16 | 264.7 | 241.8 | 
| 32 | 158.6 | 139.1 |

Again, an improvement across the board (and I think it even beats the hybrid version for these cases).

These times aren't as good as your runs on Crystal -- are the cores on that machine just faster?

**Comment by ronawho** [comment]
> These times aren't as good as your runs on Crystal -- are the cores on that machine just faster?

Hmm, that's interesting:

| Machine | CPU                                | MEM              |
| ------- | ---------------------------------- | ---------------  |
| Crystal | (2) 18-core 2.1 GHz Broadwell CPUs | 128 GB DDR4-2400 |
| Swan    | (2) 24-core 2.2 GHz Broadwell CPUs | 128 GB DDR4-2400 ||

So Swan should have slightly faster cores and more of them. I'm not sure why Swan seems to run slower, I would expect it to be the same or a little faster. (I wouldn't be surprised if they ran at the same speed since we're almost certainly saturating memory bandwidth and the Aries BTE)

Maybe split off a new issue for this, and we can check if the reference performance is also a little slower on swan. (I'm not too worried about this ATM, but it is curious)

**Comment by ronawho** [comment]
I see a little bit better performance by not oversubscribing tasks for the comm. I think the oversubscription hurts cache and we're already getting some hardware-level com/compute overlap (similar to details section in https://github.com/npadmana/DistributedFFT/issues/15#issuecomment-532854447)

I'm running full configs for that.

**Comment by ronawho** [comment]
Runs with switching comm to not oversub -- removing the `*2` from https://github.com/npadmana/DistributedFFT/blob/9c5540ca51b09c937aa6f411434193faa5fcac9d/src/DistributedFFT.chpl#L378

Size D:
=======

| nodes | ref MPI | master | pencilFFT-2 | FFT-2-comm-yz |
| ----- | ------- | ------ | ----------- | ------------- |
|   1   | 262.2s  | 157.7s | 143.5s      | 146.2s        |
|   2   | 156.5s  |  84.8s |  92.8s      |  91.1s        |
|   4   |  93.8s  |  60.8s |  70.1s      |  66.8s        |
|   8   |  51.0s  |  36.5s |  40.6s      |  40.2s        |
|  16   |  25.4s  |  20.9s |  24.8s      |  24.0s        |
|  32   |  14.1s  |  14.5s |  13.4s      |  11.6s        |
|  64   |  10.4s  |   9.6s |   7.0s      |   6.4s        |
| 128   |   5.2s  |   7.4s |   3.8s      |   3.5s        |
| 256   |   2.8s  |   6.1s |   2.1s      |   2.0s        |
| 512   |   1.6s  |   5.6s |   1.3s      |   1.2s        |


Size E:
=======

| nodes | ref MPI | master | pencilFFT-2 | FFT-2-comm-yz |
| ----- | ------- | ------ | ----------- | ------------- |
|   8   | 510.7s  | 290.8s | 474.9s      | 477.1s        |
|  16   | 240.1s  | 163.1s | 245.5s      | 258.7s        |
|  32   | 115.8s  |  99.3s | 144.0s      | 139.9s        |
|  64   |  59.0s  |  63.6s |  97.0s      |  90.2s        |
| 128   |  51.8s  |  42.5s |  50.7s      |  46.4s        |
| 256   |  25.7s  |  29.9s |  28.2s      |  24.2s        |
| 512   |  12.6s  |  25.0s |  14.2s      |  13.0s        |


Size F:
=======

| nodes | ref MPI | master | pencilFFT-2 | FFT-2-comm-yz |
| ----- | ------- | ------ | ----------- | ------------- |
|  64   | 974.0s  | OOM    | 623.2s      | 650.9s        |
| 128   | 457.0s  | 316.9s | 436.3s      | 395.7s        |
| 256   | 266.7s  | 193.2s | 243.8s      | 205.5s        |
| 512   | 133.7s  | 138.1s | 123.1s      | 110.3s        |

This definitely closes some of the gaps with master

**Comment by ronawho** [comment]
> Again, an improvement across the board (and I think it even beats the hybrid version for these cases).

Overall, hybrid seems better to me for larger problem sizes at lower node counts.

hybrid-yz vs comm-yz:
=====================


| nodes | master | FFT-2-comm-yz | FFT-2-hybrid-yz |
| ----- | ------ | ------------- | --------------- |
|   1   | 157.7s | 146.2s        | 132.3s          |
|   2   |  84.8s |  91.1s        |  89.2s          |
|   4   |  60.8s |  66.8s        |  68.5s          |
|   8   |  36.5s |  40.2s        |  40.6s          |
|  16   |  20.9s |  24.0s        |  22.6s          |
|  32   |  14.5s |  11.6s        |  12.3s          |
|  64   |   9.6s |   6.4s        |   7.0s          |
| 128   |   7.4s |   3.5s        |   3.8s          |
| 256   |   6.1s |   2.0s        |   2.1s          |
| 512   |   5.6s |   1.2s        |   1.3s          |


Size E:
=======

| nodes | master | FFT-2-comm-yz | FFT-2-hybrid-yz |
| ----- | ------ | ------------- | --------------- |
|   8   | 290.8s | 477.1s        |  397.3s         |
|  16   | 163.1s | 258.7s        |  213.6s         |
|  32   |  99.3s | 139.9s        |  120.6s         |
|  64   |  63.6s |  90.2s        |   73.5s         |
| 128   |  42.5s |  46.4s        |   50.7s         |
| 256   |  29.9s |  24.2s        |   28.1s         |
| 512   |  25.0s |  13.0s        |   14.2s         |


Size F:
=======

| nodes | master | FFT-2-comm-yz | FFT-2-hybrid-yz |
| ----- | ------ | ------------- | --------------- |
|  64   | OOM    | 650.9s        | 539.0s          |
| 128   | 316.9s | 395.7s        | 333.8s          |
| 256   | 193.2s | 205.5s        | 242.2s          |
| 512   | 138.1s | 110.3s        | 122.8s          |

**Comment by ronawho** [comment]
If we time out for additional tuning, I think I'd go with the hybrid-yz strategy. 

I'm happy to do some more runs tomorrow, but I'm timing out for tuning/investigating. I'd be very happy with our results hybrid-yz times against the reference. I think we only lose to the reference by a little for size E >32 nodes and are otherwise beating the reference.

**Comment by npadmana** [comment]
@ronawho -- last tune-up f1c45aacf57910cd7

I've blocked the non-communicating FFTW calls, so it's a ~merge of the hybrid and comm strategies. I'm seeing some non-trivial improvements with oversubscribing in the comm stage on swan, so I added a switch `--overSubscribe`. 

Any chance you run timings on this, and regenerate plots....? I'll resubmit the abstract with new plots.

**Comment by ronawho** [comment]
Yup, running now.

**Comment by npadmana** [comment]
@ronawho -- FYI. I just realized I set `overSubscribe` to true by default. So to turn it off, you need `overSubscribe=false`. Just in case you didn't read the code....  :)

**Comment by ronawho** [comment]
Yeah, I'm running both ways

**Comment by npadmana** [comment]
On swan, case D. The numbers are the mean and standard deviation of 5 runs.

|nodes| `overSub=true`|`overSub=false`|
|---|---|---|
|8| 34.403(0.495575) | 36.3621(1.16759)|
|16| 21.7992(0.996377)| 22.6584(0.954364) |
|32| 13.1119(0.148766)| 12.9865(0.237503) |

**Comment by npadmana** [comment]
I think I figured out why I was seeing improvements with `overSubscribe` previously. My PBS scripts had a leftover `ncpus=44` in them, which I think meant that I wasn't getting all cores. Removing that, I get pretty consistent numbers with both.

**Comment by ronawho** [comment]
Ah interesting, could also explain why your numbers were worse than crystal, but we should double-check that. 

Runs and graphs should be done in an hour.

**Comment by npadmana** [comment]
@ronawho — I wonder, given that we are likely memory and network bandwidth
limited, whether we’d win by using hyperthreading.

Or does the over subscription effectively do this?

**Comment by ronawho** [comment]
Results for f1c45aa, but with oversub set to false (I see better performance at higher node counts). I'll email the pdfs

Size D:
=======

| nodes | ref MPI | master | FFT-2-comm-yz |
| ----- | ------- | ------ | ------------- |
|   1   | 262.2s  | 157.7s | 160.8s        |
|   2   | 156.5s  |  84.8s |  94.7s        |
|   4   |  93.8s  |  60.8s |  61.7s        |
|   8   |  51.0s  |  36.5s |  37.4s        |
|  16   |  25.4s  |  20.9s |  21.3s        |
|  32   |  14.1s  |  14.5s |  11.6s        |
|  64   |  10.4s  |   9.6s |   6.6s        |
| 128   |   5.2s  |   7.4s |   3.9s        |
| 256   |   2.8s  |   6.1s |   2.3s        |
| 512   |   1.6s  |   5.6s |   1.5s        |


Size E:
=======

| nodes | ref MPI | master | FFT-2-comm-yz |
| ----- | ------- | ------ | ------------- |
|   8   | 510.7s  | 290.8s | 366.4s        |
|  16   | 240.1s  | 163.1s | 201.6s        |
|  32   | 115.8s  |  99.3s | 105.9s        |
|  64   |  59.0s  |  63.6s |  59.4s        |
| 128   |  51.8s  |  42.5s |  31.0s        |
| 256   |  25.7s  |  29.9s |  18.1s        |
| 512   |  12.6s  |  25.0s |  10.0s        |


Size F:
=======

| nodes | ref MPI | master | FFT-2-comm-yz |
| ----- | ------- | ------ | ------------- |
|  64   | 974.0s  | OOM    | 484.0s        |
| 128   | 457.0s  | 316.9s | 251.4s        |
| 256   | 266.7s  | 193.2s | 149.5s        |
| 512   | 133.7s  | 138.1s |  78.2s        |

**Comment by ronawho** [comment]
Fresh runs for reference vs f1c45aa with oversub set to false, using Chapel 1.20 release. Best of 3 trials:

![npb-ft-D-perf](https://user-images.githubusercontent.com/1588337/65925210-dac24680-e3bd-11e9-87f4-72c1ee9b318f.png)

![npb-ft-F-perf](https://user-images.githubusercontent.com/1588337/65925219-e150be00-e3bd-11e9-8521-55e1720a0cd9.png)

![npb-ft-E-perf](https://user-images.githubusercontent.com/1588337/65925213-ddbd3700-e3bd-11e9-87df-fb22b9ff824d.png)

Performance looks pretty awesome!

 It would be great to do some code cleanup and I think there are still some places master performed better, but I think these results are a huge improvement over what we had before.

**Comment by ronawho** [comment]
>  I wonder, given that we are likely memory and network bandwidth limited, whether we’d win by using hyperthreading.

We really only see hyperthreading benefit for I/O or fine-grained comm heavy benchmarks. Adding hyperthreads usually hurts or is a noop for compute or bandwith bound benchmarks, and we're doing enough bulk comm that I don't think we'll see any performance benefit from adding more threads. That said, it's worth trying -- we can do some runs with `CHPL_RT_NUM_THREADS_PER_LOCALE=MAX_LOGICAL`

**Comment by npadmana** [comment]
Those perf graphs look great!

One caveat — the ref MPI version is limited to 32 cores per node; it would
be interesting to see how much the Chapel performance gets worse if we
limit the number of cores.

**Comment by npadmana** [comment]
A related comment is that my swan runs still trail your Crystal runs — so
that would suggest weak scaling with the number of cores.

I wonder if I’m using eg a different hugepages.

**Comment by ronawho** [comment]
> One caveat — the ref MPI version is limited to 32 cores per node; it would
be interesting to see how much the Chapel performance gets worse if we
limit the number of cores.

Yeah, that's true. Last time I ran, I did not see performance loss from limiting chapel to 32 cores (suspecting that we're hitting memory and bulk transfer bandwidth already) 


> A related comment is that my swan runs still trail your Crystal runs — so that would suggest weak scaling with the number of cores.

> I wonder if I’m using eg a different hugepages.

Last I looked at your logs we were both using 16M hugepages. 


Tomorrow, I'll do some size E runs on swan and crystal for both their natural number of cores and limiting both to 32.
